cpu/nrf51/radio/nrf51_ble/nrf51_ble.c

```diff
@@ -0,0 +1,630 @@
+ */
+static state_t _tx_prestate;
+
+/**
+ * @brief    Double receive buffers
+ */
+static ble_pkt_t _rx_buf[2];
+
+/**
+ * @brief   BLE Address of the device
+ */
+static uint32_t _access_addr = BLE_DEFAULT_ACCESS_ADDR;
+
+/**
+ * @brief   Pointer to the free receive buffer
+ */
+static volatile int _rx_next = 0;
+
+
+/* Switch the BLE radio state to idle */
+static void _switch_to_idle(void)
+{
+    // DEBUG("nrf51_ble: switch_to_idle\n");
+
+    /* switch to idle state */
+    NRF_RADIO->EVENTS_DISABLED = 0;
+    NRF_RADIO->TASKS_DISABLE = 1;
+
+    while (NRF_RADIO->EVENTS_DISABLED == 0);
+
+    _state = STATE_IDLE;
+}
+
+/* Switch the BLE radio state to receive */
+static void _switch_to_rx(void)
+{
+    // DEBUG("nrf51_ble: switch_to_rx\n");
+
+    /* set pointer to receive buffer */
+    NRF_RADIO->PACKETPTR = (uint32_t) & (_rx_buf[_rx_next]);
+
+    /* switch int RX mode */
+    NRF_RADIO->TASKS_RXEN = 1;
+    _state = STATE_RX;
+}
+
+/* Get the BLE Access Address of the interface */
+int _get_access_address(uint8_t *val, size_t max_len)
+{
+    /* check parameters */
+    if (max_len < BLE_ACCESS_ADDR_LEN) {
+        return -EOVERFLOW;
+    }
+
+    /* get address */
+    val[0] = (uint8_t)(_access_addr >> 24);
+    val[1] = (uint8_t)(_access_addr >> 16);
+    val[2] = (uint8_t)(_access_addr >> 8);
+    val[3] = (uint8_t)(_access_addr);
+    return BLE_ACCESS_ADDR_LEN;
+}
+
+/* Set the BLE Access Address of the interface */
+int _set_access_address(uint8_t *val, size_t len)
+{
+    int is_rx = 0;
+
+    /* check parameters */
+    if (len != BLE_ACCESS_ADDR_LEN) {
+        return -EINVAL;
+    }
+
+    /* keep track of state */
+    while (_state == STATE_TX);
+
+    if (_state == STATE_RX) {
+        is_rx = 1;
+        _switch_to_idle();
+    }
+
+    /* set BLE access address */
+    _access_addr = (((uint32_t)val[0]) << 24) | (((uint32_t)val[1]) << 16) |
+                   (((uint32_t)val[2]) << 8) | val[3];
+
+    NRF_RADIO->PREFIX0 = ((_access_addr >> 24) & 0x000000FF);
+    NRF_RADIO->BASE0   = ((_access_addr << 8) & 0xFFFFFF00);
+
+    if (is_rx) {
+        _switch_to_rx();
+    }
+
+    return BLE_ACCESS_ADDR_LEN;
+}
+
+/* Get the BLE radio state */
+int _get_state(uint8_t *val, size_t max_len)
+{
+    ng_netopt_state_t state;
+
+    /* check parameters */
+    if (max_len < sizeof(ng_netopt_state_t)) {
+        return -EOVERFLOW;
+    }
+
+    switch (_state) {
+        case STATE_OFF:
+            state = NG_NETOPT_STATE_OFF;
+            break;
+
+        case STATE_IDLE:
+            state = NG_NETOPT_STATE_SLEEP;
+            break;
+
+        case STATE_RX:
+            state = NG_NETOPT_STATE_IDLE;
+            break;
+
+        case STATE_TX:
+            state = NG_NETOPT_STATE_TX;
+            break;
+
+        default:
+            return -ECANCELED;
+    }
+
+    memcpy(val, &state, sizeof(ng_netopt_state_t));
+    return sizeof(ng_netopt_state_t);
```

Comment:
Do we need that return value? Otherwise one could just return the state and discard the memcpys in 212.